black.py

```diff
@@ -454,7 +454,9 @@ def format_file_in_place(
         return False
 
     if write_back == write_back.YES:
-        with open(src, "w", encoding=src_buffer.encoding) as f:
+        with open(src, encoding=src_buffer.encoding, newline="") as src_buffer:
```

Comment:
You don't have to open it again, that's slow. You already have `src_contents` that you can check. Suggestion:
```py3
windows_newlines = src_contents.split("\n", 1)[0][-1:] == "\r"
newline = "\r\n" if windows_newlines else "\n"
```

Reply:
I'm not sure if opening a file for reading one more time right after reading through it is a huge performance hit, but if it turns out to be the case, we can reproduce the 10 or so lines of code in `tokenize.open()` to make sure we only do it once.

Reply:
Right. OK, leave as is but open it in `"rb"` so it doesn't need to decode anything. The first line must be readable anyway since if it's not UTF8 (or ASCII on Python 2) it needs the `# -*- coding: X -*-` comment.

And as a total nit: instead of `b"\r\n" in readline` you can do `readline[-2:] == b"\r\n"` which is marginally faster still.